File: round-c/circuit_board/circuit_board_large.py
from collections import Counter, defaultdict
from math import ceil
from bisect import bisect_left, bisect_right
from random import randint


class Solution():
    def compute_logs(self, n):
        logs = [0, 0]
        for i in range(2, n + 1):
            logs.append(logs[i // 2] + 1)
        return logs

    def build(self, arr, l):
        min_table = [[0] * 10 for _ in range(l)]
        max_table = [[0] * 10 for _ in range(l)]
        for i in range(l):
            max_table[i][0] = arr[i]
            min_table[i][0] = arr[i]

        for j in range(1, 11):
            for i in range(l - (1 << j) + 1):
                max_table[i][j] = max(max_table[i][j - 1], max_table[i + (1 << (j - 1))][j - 1])
                min_table[i][j] = min(min_table[i][j - 1], min_table[i + (1 << (j - 1))][j - 1])
        return max_table, min_table

    def get_min(self, l, r, table, logs):
        j = logs[r - l + 1]
        return min(table[l][j], table[r - 2 ** j + 1][j])

    def get_max(self, l, r, table, logs):
        j = logs[r - l + 1]
        return max(table[l][j], table[r - 2 ** j + 1][j])

    def max_rentagle(self, histogram):
        stack = list()

        max_area = 0

        index = 0
        while index < len(histogram):
            if (not stack) or (histogram[stack[-1]] <= histogram[index]):
                stack.append(index)
                index += 1

            else:
                top_of_stack = stack.pop()

                area = (histogram[top_of_stack] *
                        ((index - stack[-1] - 1)
                         if stack else index))

                max_area = max(max_area, area)

        while stack:
            top_of_stack = stack.pop()

            area = (histogram[top_of_stack] *
                    ((index - stack[-1] - 1)
                     if stack else index))

            max_area = max(max_area, area)

        return max_area

    def search(self, max_table, min_table, logs, j, k, n, c):
        s, e = j, n
        # print('search', j)
        while s < e - 1:
            m = (s + e) // 2
            max_val = self.get_max(j, m, max_table, logs)
            min_val = self.get_min(j, m, min_table, logs)
            # print('max', j, m, max_val)
            # print('min', min_val)
            if abs(max_val - c) <= k and abs(min_val - c) <= k:
                s = m
            else:
                e = m - 1
        # print(s)
        # print('')
        return s - j + 1

    def solve(self):
        R, C, K = [int(e) for e in input().split()]
        rows = []
        for _ in range(R):
            rows.append([int(e) for e in input().split()])

        mem = [[0] * C for _ in range(R)]
        logs = self.compute_logs(C)
        for i, row in enumerate(rows):
            max_table, min_table = self.build(row, C)
            print(max_table);
            for j in range(C):
                mem[i][j] = self.search(max_table, min_table, logs, j, K, C, row[j])

        # print(mem)
        ans = 0
        for i in range(C):
            bars = [mem[j][i] for j in range(R)]
            ans = max(ans, self.max_rentagle(bars))

        return ans


for t in range(int(input())):
    ans = Solution().solve()
    print('Case #{}: {}'.format(t + 1, ans))

File: round-c/circuit_board/circuit_board_large.cpp
#include <iostream>
#include <map>
#include <vector>
#include <tuple>
#include <array>
#include <stack>
#include <math.h>
#include <string>
#include <sstream>
#include <algorithm>
#include <iterator>

using namespace std;

void print(vector<int> &arr) {
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

struct Tables {
    vector<vector<int>> max_table;
    vector<vector<int>> min_table;
};

vector<int> compute_logs(int n) {
    vector<int> logs = {0, 0};
    for (int i = 2; i < n + 1; i++) {
        logs.push_back(logs[i / 2] + 1);
    }
    return logs;
}

Tables build(vector<int> &arr, int l) {
    vector<vector<int>> min_table(l, vector<int> (10, 0));
    vector<vector<int>> max_table(l, vector<int> (10, 0));

    for (int i = 0; i < l; i++) {
        max_table[i][0] = arr[i];
        min_table[i][0] = arr[i];
    }

    for (int j = 1; j < 11; j++) {
        for (int i = 0; i < l - (1 << j) + 1; i++) {
            max_table[i][j] = max(max_table[i][j - 1], max_table[i + (1 << (j - 1))][j - 1]);
            min_table[i][j] = min(min_table[i][j - 1], min_table[i + (1 << (j - 1))][j - 1]);
        }
    }

    return {max_table, min_table};
}

int get_min(int l, int r, vector<vector<int>> &table, vector<int> &logs) {
    int j = logs[r - l + 1];
    return min(table[l][j], table[r - (1 << j) + 1][j]);
}

int get_max(int l, int r, vector<vector<int>> &table, vector<int> &logs) {
    int j = logs[r - l + 1];
    return max(table[l][j], table[r - (1 << j) + 1][j]);
}

int max_rentagle(int hist[], int n)
{
    stack<int> s;

    int max_area = 0;
    int tp;
    int area_with_top;

    int i = 0;
    while (i < n)
    {
        if (s.empty() || hist[s.top()] <= hist[i])
            s.push(i++);
        else
        {
            tp = s.top();
            s.pop();

            area_with_top = hist[tp] * (s.empty() ? i :
                                        i - s.top() - 1);

            if (max_area < area_with_top)
                max_area = area_with_top;
        }
    }

    while (!s.empty())
    {
        tp = s.top();
        s.pop();
        area_with_top = hist[tp] * (s.empty() ? i :
                                    i - s.top() - 1);

        if (max_area < area_with_top)
            max_area = area_with_top;
    }

    return max_area;
}

int search(vector<vector<int>> &max_table, vector<vector<int>> &min_table, vector<int> &logs, int j, int k, int n, int c) {
    int s = j, e = n;
    while (s < e - 1) {
        int m = (s + e) / 2;
        int max_val = get_max(j, m, max_table, logs);
        int min_val = get_min(j, m, min_table, logs);

        if (max_val - min_val <= k) {
            s = m;
        } else {
            e = m;
        }
    }
    return s - j + 1;
}

int solve() {
    int R, C, K;
    cin >> R >> C >> K;

    vector<vector<int>> rows;
    for (int i = 0; i < R; i++) {
        vector<int> v;
        for (int j = 0; j < C; j++) {
            int n;
            cin >> n;
            v.push_back(n);
        }
        rows.push_back(v);
    }

    vector<vector<int>> mem(R, vector<int>(C, 0));
    vector<int> logs = compute_logs(C);
    for (int i = 0; i < R; i++) {
        Tables t = build(rows[i], C);
        vector<vector<int>> max_table = t.max_table, min_table = t.min_table;
        for (int j = 0; j < C; j++) {
            mem[i][j] = search(max_table, min_table, logs, j, K, C, rows[i][j]);
        }
    }

    int ans = 0;
    for (int i = 0; i < C; i++) {
        int bars[R];
        for(int j = 0; j < R; j++) {
            bars[j] = mem[j][i];
        }
        ans = max(ans, max_rentagle(bars, R));
    }
    return ans;
}

int main() {
    int tc;
    cin >> tc;
    for (int tt = 1; tt <= tc; tt++) {
        int ans = solve();
        cout << "Case #" << tt << ": " << ans << endl;
    }
    return 0;
}
